### ISSUE TITLE
react-native support

### ISSUE THREAD

**Issue by dyegolara** [nbd-wtf/nostr-tools#80]
I have tried to polyfill my react-native app so it could use the `crypto` native module, but then I realized that this library is meant to be used by clients, a lot of which doesn't support node's native `crypto` module.

May I suggest using some library instead, like https://github.com/entronad/crypto-es ?
Do you think it would be possible?

**Comment by aussedatlo** [comment]
With Expo and react-native, you can patch the new version of `@noble/secp256k1` to use `expo-random` instead of node's native `crypto` module.

It's not the best solution but it work.

**Comment by fiatjaf** [comment]
I forgot to answer this, but that was fine because I didn't have an answer, and @aussedatlo has provided us with a potential solution here that I don't understand either.

@koalasat any insight here?

**Comment by KoalaSat** [comment]
Yes, I hacked it for Nostros by forking the `cipher-base` library and changing changing this dependency https://github.com/crypto-browserify/cipher-base/compare/master...KoalaSat:cipher-base:master

At that point I'm just downloading it on my package https://github.com/KoalaSat/nostros/blob/main/package.json#L22

As a little bit of background, the issue is that `nostr-tools` depends on an ECMA version that Android still does not implements (can't remember the number it was like 19 vs 21 or something)

**Comment by KoalaSat** [comment]
you might also have problems with Buffer, let me know if it's your case @dyegolara

**Comment by fiatjaf** [comment]
In the newest nostr-tools version there is no usage of Buffer anymore.

**Comment by KoalaSat** [comment]
I wasn't aware of that then maybe I'll try of get rid of the hack for Nostros

**Comment by facinick** [comment]
what's the status of this issue, does nostr-tools support react native now?

**Comment by Eosxx** [comment]
Try this.

```typescript
import * as secp256k1 from "@noble/secp256k1";
import { randomBytes  } from 'react-native-randombytes'

secp256k1.utils.randomBytes = function (bytesLength  = 32) {
  return Uint8Array.from(randomBytes(bytesLength));
}
```

**Comment by earonesty** [comment]
@Eosxx even with randomBytes working, crypto.subtle isn't going to work on RN.  not sure how this is a solution.   need a polyfill for subtle.

**Comment by earonesty** [comment]
looks like other projects just don't use the libs from nostr-tools instead.... https://github.com/KoalaSat/nostros/blob/main/frontend/lib/nostr/Nip04/index.ts.  im having some luck with isomorphic-webcrypto.   everything works but the randomBytes.   going to patch it up soon, and then release a new version (maintainer unresponsive).   don't like "rolling my own".  nostr-tools has more eyeballs.

**Comment by JoaoLobo94** [comment]
Any news on this? trying to implement a client with expo react-native and functions such as getSignature and generatePrivateKey do not seem to work.

I have had a quick fix for generatePrivate key such as

```
async function generatePrivateKey(): Promise<string> {
  const privateKeyBytes = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, Math.random().toString());
  return privateKeyBytes.substring(0, 64);
}
```

Any tips on getting around this?

**Comment by earonesty** [comment]
i use this: https://github.com/earonesty/isomorphic-webcrypto

works fine

**Comment by JoaoLobo94** [comment]
So you are currently working on a PR to add react native support?

Or the solution you say is to use those libraries and emulate the functions from nostr tools?

**Comment by thevikas** [comment]
I am getting `Property TextDecoder doesn't exists` on my RN usage. Any way around it? any polyfills that work to implement this method in RN?

**Comment by earonesty** [comment]
https://www.npmjs.com/package/text-encoding-polyfill

**Comment by thevikas** [comment]
Is there any react native proj using nostr-tools? the getRandomValues is also undefined.

**Comment by minibits-cash** [comment]
I use nostr-tools in my RN project the following way: 

1. Install react-native-quick-crypto following their readme. It exposes part of the node crypto functions using C++ implementation and JSI bridge - fast and without the need for shim.js
2. There are still some functions missing that are used by nostr-tools, notably crypto.subtle used e.g. for encrypting NIP04 messages. However encryption/decryption can be easily implemented using examples in nip04 docs that does not use subtle, just replace node crypto functions with their Quickcrypto implementation.

**Comment by SamSamskies** [comment]
I used [react-native-get-random-values](https://www.npmjs.com/package/react-native-get-random-values) to polyfill `crypto.getRandomValues`. It's recommended in the README for [@noble/curves](https://github.com/paulmillr/noble-curves#usage). Haven't tested much, but signing events and generating private keys are now working for me.

**Comment by SamSamskies** [comment]
Using a SimplePool instance doesn't work for me in react native though. I haven't been able to figure out why. I've tried both `pool.publish` and `pool.get` and both don't work. Promises all get rejected. Works fine if I use `relay.publish` and `relay.get`.

**Comment by gasparnd** [comment]
Someone can use nostr-tools in React Native? I have the next problems:
Error: crypto.getRandomValues must be defined
Error: crypto.getRandomValues must be defined

**Comment by gasparnd** [comment]
> With Expo and react-native, you can patch the new version of `@noble/secp256k1` to use `expo-random` instead of node's native `crypto` module.
> 
> It's not the best solution but it work.

Can your share the code please, i have a bare react native app with Expo modules and it doesn't work for me

**Comment by SamSamskies** [comment]
> > With Expo and react-native, you can patch the new version of `@noble/secp256k1` to use `expo-random` instead of node's native `crypto` module.
> > It's not the best solution but it work.
> 
> Can your share the code please, i have a bare react native app with Expo modules and it doesn't work for me

Try using `react-native-get-random-values`. That's I used for the Wavlake mobile app.

https://github.com/wavlake/mobile/blob/467b0824c5adde4da556e86e478ee9a6a1b4cb61/utils/nostr.ts#L4-L5
